staging/src/k8s.io/component-base/logs/logrotation/logrotation.go

```diff
@@ -0,0 +1,233 @@
+	}
+
+	return w, nil
+}
+
+// write func to satisfy io.writer interface
+func (w *RotationFile) Write(p []byte) (n int, err error) {
+	w.mut.Lock()
+	defer w.mut.Unlock()
+
+	n, err = w.file.Write(p)
+	if err != nil {
+		return 0, err
+	}
+
+	if !w.isFlushed {
+		err = w.file.Sync()
+		if err != nil {
+			return 0, err
+		}
+		w.isFlushed = true
+	}
+
+	if w.maxSize > 0 {
+		w.currentSize += int64(len(p))
+
+		// if file size over maxsize rotate the log file
+		if w.currentSize >= w.maxSize {
+			// Explicitly call file.Sync() to ensure data is written to disk
+			err = w.file.Sync()
+			if err != nil {
+				return 0, err
+			}
+			err = w.rotate()
+			if err != nil {
+				return 0, err
+			}
+		}
+	}
+
+	return n, nil
+}
+
+func (w *RotationFile) rotate() error {
+	// Get the file extension
+	ext := filepath.Ext(w.filePath)
+
+	// Remove the extension from the filename
+	pathWithoutExt := strings.TrimSuffix(w.filePath, ext)
+
+	rotateFilePath := pathWithoutExt + "-" + time.Now().Format(timeLayout) + ext
+
+	if w.filePath == rotateFilePath {
+		return nil
+	}
+
+	err := w.file.Close()
+	if err != nil {
+		return err
+	}
+
+	err = os.Rename(w.filePath, rotateFilePath)
+	if err != nil {
+		return err
+	}
+
+	w.file, err = os.OpenFile(w.filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+
+	w.currentSize = 0
+
+	if w.maxAge > 0 {
+		go func() {
+			err = w.clean(pathWithoutExt, ext)
+		}()
+	}
+
+	return nil
+}
+
+// Clean up the old log files in the format of
+// <basename>-<timestamp><ext>
+// This should be safe enough to avoid false deletion
+// This will work for multiple restarts of the same program
+func (w *RotationFile) clean(pathWithoutExt string, ext string) error {
+	ageTime := time.Now().Add(-w.maxAge)
+
+	directory := filepath.Dir(pathWithoutExt)
+	basename := filepath.Base(pathWithoutExt) + "-"
+
+	dir, err := os.ReadDir(directory)
+	if err != nil {
+		return err
+	}
+
+	err = nil
+	for _, v := range dir {
+		if strings.HasPrefix(v.Name(), basename) && strings.HasSuffix(v.Name(), ext) {
+			// Remove the prefix and suffix
+			trimmed := strings.TrimPrefix(v.Name(), basename)
+			trimmed = strings.TrimSuffix(trimmed, ext)
+
+			_, err = time.Parse(timeLayout, trimmed)
+			if err == nil {
+				info, errInfo := v.Info()
+				if errInfo != nil {
+					err = errInfo
+					// Ignore the error while continue with the next clenup
+					continue
+				}
+
+				if ageTime.After(info.ModTime()) {
+					err = os.Remove(filepath.Join(directory, v.Name()))
+					if err != nil {
+						// Ignore the error while continue with the next clenup
+						continue
+					}
+				}
+			}
+
+		}
+	}
+
+	return err
+}
+
+func (w *RotationFile) Close() error {
+	w.mut.Lock()
+	defer w.mut.Unlock()
+
+	// Explicitly call file.Sync() to ensure data is written to disk
+	err := w.file.Sync()
+	if err != nil {
+		return err
+	}
+
+	err = w.file.Close()
+	if err != nil {
+		return err
+	}
+
+	if w.stop != nil {
+		close(w.stop)
+	}
```

Comment:
Let's also block until that goroutine has stopped.

Wouldn't it be simpler to check the current time during each `Write` instead of running a ticker in a goroutine?

Reply:
I think it is a god idea, let me switch to that way